File: mac/PabloCompanion/ViewModels/TranscriptionViewModel.swift
import CompanionSessionCore
import Foundation
import os

// MARK: - DecryptedPCMPaths

/// Resolved paths after decrypting encrypted PCM sidecars for upload.
private struct DecryptedPCMPaths {
    let micPath: String
    let systemPath: String?
    let tempFiles: [URL]
}

// MARK: - TranscriptionState

enum TranscriptionState: Sendable {
    case running
    case done(transcript: String)
    case pendingUpload(transcript: String)
    case failed(message: String)

    var transcript: String? {
        switch self {
        case let .done(text), let .pendingUpload(text): text
        default: nil
        }
    }

    var isPendingUpload: Bool {
        if case .pendingUpload = self { return true }
        return false
    }
}

// MARK: - TranscriptionViewModel

/// Orchestrates cloud-based transcription after a session ends.
///
/// Flow:
///   1. `transcribeIfNeeded(_:)` — checks auto-transcribe setting, uploads audio
///   2. Uploads mic + system audio to the backend for server-side transcription
///   3. On upload failure: the entry stays queued in `PendingAudioUploadStore`
///   4. `retryPendingAudioUploads()` — drains it on launch and every 5 minutes
@MainActor
@Observable
final class TranscriptionViewModel {
    // MARK: - State

    /// Transcription state keyed by recording ID.
    var states: [UUID: TranscriptionState] = [:]

    /// Number of transcripts waiting to be uploaded.
    var pendingUploadCount = 0

    var errorMessage: String?
    var showError = false

    // MARK: - Dependencies

    var backendURL = "" {
        didSet {
            if URLValidator.validateScheme(backendURL) == nil {
                let token = apiClient.getToken
                let onAuthRejected = apiClient.onAuthRejected
                apiClient = APIClient(baseURL: backendURL)
                apiClient.getToken = token
                apiClient.onAuthRejected = onAuthRejected
            }
        }
    }

    // MARK: - Private

    private var apiClient = APIClient()
    private var audioStore = PendingAudioUploadStore(
        directory: AppPaths.pendingAudioUploads,
        makeEncryptor: { RecordingEncryptor(userEmail: $0) },
        logSubsystem: AppConstants.appBundleID
    )
    private let logger = Logger(subsystem: AppConstants.appBundleID, category: "TranscriptionViewModel")

    /// The signed-in user's email, used to scope encryption keys.
    var userEmail: String? {
        didSet {
            audioStore.userEmail = userEmail
        }
    }

    // Exponential backoff for audio-upload retries — parity with Windows
    // (TranscriptionViewModel.cs:30-32).
    private let audioBaseBackoffSeconds: Double = 300
    private let audioMaxBackoffSeconds: Double = 14400
    private let audioMaxAutoRetries = 10

    /// Rate assumed for entries queued before `sampleRate` was persisted. Raw
    /// PCM has no header to recover the real rate from, so a legacy entry can
    /// only be guessed — 48 kHz matches what the app assumed unconditionally
    /// before the capture rate was plumbed through.
    private static let fallbackSampleRate: Double = 48000

    private var autoTranscribe: Bool {
        UserDefaults.standard.object(forKey: "autoTranscribe") as? Bool ?? true
    }

    /// Configures the API client with a token provider for authenticated
    /// requests and an optional handler for server-side session rejection.
    func configureAuth(
        getToken: @escaping @Sendable () async throws -> String,
        onAuthRejected: ((Bool) -> Void)? = nil
    ) {
        apiClient.getToken = getToken
        apiClient.onAuthRejected = onAuthRejected
    }

    // MARK: - Public API

    /// Triggers cloud transcription for a recording if auto-transcribe is enabled.
    func transcribeIfNeeded(_ recording: LocalRecording, sessionId: String? = nil) {
        guard autoTranscribe else { return }
        guard recording.micPCMFileURL != nil else {
            logger.info("Skipping transcription: no PCM sidecar")
            return
        }
        guard let sessionId else {
            logger.info("Skipping cloud transcription: no session ID")
            return
        }
        Task { await uploadAudioToBackend(recording, sessionId: sessionId) }
    }

    /// Uploads audio files for a set of recordings to the backend for server-side transcription.
    func uploadAudioSegments(_ recordings: [LocalRecording], sessionId: String) async {
        guard autoTranscribe else { return }
        let viable = recordings.filter { $0.micPCMFileURL != nil }
        guard !viable.isEmpty else { return }

        // For cloud upload, use the last recording's audio files (multi-segment will be
        // concatenated on the backend in a future iteration).
        if let recording = viable.last {
            await uploadAudioToBackend(recording, sessionId: sessionId)
        }
    }

    /// Uploads the therapist (mic) and client (system) audio to the backend.
    ///
    /// Persists the upload intent to `PendingAudioUploadStore` BEFORE the
    /// network call so a process crash, sign-out, or network outage leaves
    /// a recovery anchor on disk that `retryPendingAudioUploads` can drain
    /// on the next launch. Mirrors Windows
    /// `TranscriptionViewModel.UploadAudioAsync` (cs:70-115).
    private func uploadAudioToBackend(_ recording: LocalRecording, sessionId: String) async {
        guard let micURL = recording.micPCMFileURL else {
            states[recording.id] = .failed(message: "No mic audio file available")
            return
        }

        // Enqueue BEFORE any network call. Idempotent — re-adding the same
        // sessionId preserves `createdAt` and `retryCount`.
        audioStore.add(
            sessionId: sessionId,
            micPath: micURL.path,
            systemPath: recording.systemPCMFileURL?.path,
            isEncrypted: recording.isEncrypted,
            sampleRate: recording.sampleRate
        )

        // Cheap read-only liveness probe before moving the audio. If the
        // server-side session has already idled out, the upload can only 401 —
        // surface the re-auth flow now (verifySessionAlive fires it) and leave
        // the entry queued. It drains via the retry loop after sign-in.
        guard await apiClient.verifySessionAlive() else {
            states[recording.id] = .failed(message: "Session expired — sign in to resume the upload")
            logger.warning("Skipping audio upload: server session is no longer active")
            return
        }

        states[recording.id] = .running
        logger.info("Uploading audio to backend for server-side transcription")

        // Same drain the retry loop uses, so the live path cannot diverge from
        // it — and so a successful upload here deletes the audio too.
        let succeeded = await coordinator.forceDrain(only: sessionId) == 1

        if succeeded {
            states[recording.id] = .done(transcript: "")
        } else {
            states[recording.id] = .failed(message: "Audio upload failed — will retry later")
        }
        pendingUploadCount = audioStore.loadAll().count
    }

    /// The tested drain: backoff ladder, retry cap, and cleanup after a
    /// confirmed upload all live in `CompanionSessionCore` so the harness and
    /// `swift test` can drive them without launching this app.
    private var coordinator: PendingAudioUploadCoordinator {
        PendingAudioUploadCoordinator(
            store: audioStore,
            policy: .init(
                baseBackoffSeconds: audioBaseBackoffSeconds,
                maxBackoffSeconds: audioMaxBackoffSeconds,
                maxAutoRetries: audioMaxAutoRetries
            ),
            upload: { entry in
                try await self.upload(entry)
            },
            cleanup: { entry in
                RecordingCleaner.removeAudio(micPath: entry.micPath, systemPath: entry.systemPath)
            },
            checkOutcome: { [apiClient] sessionId in
                // Local audio is PHI; it is deleted only once the backend has
                // produced the note, never merely because the upload was
                // accepted. A transient backend failure that once deleted the
                // audio on the ack left the session unrecoverable.
                let session = try await apiClient.fetchSession(sessionId: sessionId)
                switch session.status {
                case .pendingReview, .finalized:
                    return .noteReady
                case .failed:
                    return .failed
                default:
                    // transcribing / queued / processing / anything mid-flight
                    return .stillWorking
                }
            },
            logSubsystem: AppConstants.appBundleID
        )
    }

    /// One upload attempt. Throws on failure so the coordinator can count the
    /// retry; the `INVALID_STATUS` self-heal lives in `AudioUploadClient`.
    private func upload(_ entry: PendingAudioUploadStore.PendingAudioUpload) async throws {
        let pcm = try decryptPCMIfNeeded(
            micPath: entry.micPath,
            systemPath: entry.systemPath,
            isEncrypted: entry.isEncrypted
        )
        defer { pcm.tempFiles.forEach { RecordingEncryptor.cleanupTempFile($0) } }

        _ = try await apiClient.uploadAudioWithSelfHeal(
            sessionId: entry.sessionId,
            therapistAudioURL: URL(fileURLWithPath: pcm.micPath),
            clientAudioURL: pcm.systemPath.map { URL(fileURLWithPath: $0) },
            // The capture rate is negotiated at runtime (Bluetooth HFP can drop
            // the mic to 8/16/24 kHz), so stamp the WAV with the rate the
            // capture actually used, not a hardcoded 48 kHz.
            sampleRate: Int(entry.sampleRate ?? Self.fallbackSampleRate),
            onProgress: { _ in }
        )
    }

    /// Drain due entries. Invoked on launch, on a 5-minute timer, and after
    /// orphan adoption.
    func retryPendingAudioUploads() async {
        let coordinator = coordinator
        let drained = await coordinator.drain()
        if drained > 0 {
            logger.info("Drained \(drained) pending audio upload(s)")
        }
        // Same cadence: check whether any already-uploaded session now has its
        // note, so the audio can finally be deleted (or re-queued on failure).
        let confirmed = await coordinator.reconcile()
        if confirmed > 0 {
            logger.info("Confirmed \(confirmed) note(s); deleted local audio")
        }
        pendingUploadCount = audioStore.loadAll().count
    }

    /// Retry everything now, ignoring backoff and the retry cap. Bound to the
    /// Settings "Retry now" entry, where waiting out a ladder the user just
    /// overrode would be wrong.
    func forceRetryPendingAudioUploads() async {
        let drained = await coordinator.forceDrain()
        logger.info("Force-drained \(drained) pending audio upload(s)")
        pendingUploadCount = audioStore.loadAll().count
    }

    /// Enqueue an audio upload for a session whose recording lives on disk
    /// already (e.g. orphan adoption on launch). Idempotent — re-adding the
    /// same session preserves `createdAt` and `retryCount`.
    /// - Parameter sampleRate: `nil` for a recording adopted off disk, whose
    ///   capture rate is not recoverable from headerless PCM. The retry then
    ///   falls back to `fallbackSampleRate`.
    func enqueuePendingAudioUpload(
        sessionId: String,
        micPath: String,
        systemPath: String?,
        isEncrypted: Bool,
        sampleRate: Double? = nil
    ) {
        audioStore.add(
            sessionId: sessionId,
            micPath: micPath,
            systemPath: systemPath,
            isEncrypted: isEncrypted,
            sampleRate: sampleRate
        )
    }

    private func decryptPCMIfNeeded(
        micPath: String,
        systemPath: String?,
        isEncrypted: Bool
    ) throws -> DecryptedPCMPaths {
        var tempFiles: [URL] = []

        let resolvedMic: String
        if isEncrypted {
            let micURL = URL(fileURLWithPath: micPath)
            let tempURL = try RecordingEncryptor.decryptPCMToTempFile(at: micURL, userEmail: userEmail)
            tempFiles.append(tempURL)
            resolvedMic = tempURL.path
            logger.info("Decrypted mic PCM to temp file")
        } else {
            resolvedMic = micPath
        }

        let resolvedSystem: String?
        if isEncrypted, let systemPath {
            let systemURL = URL(fileURLWithPath: systemPath)
            let tempURL = try RecordingEncryptor.decryptPCMToTempFile(at: systemURL, userEmail: userEmail)
            tempFiles.append(tempURL)
            resolvedSystem = tempURL.path
            logger.info("Decrypted system PCM to temp file")
        } else {
            resolvedSystem = systemPath
        }

        return DecryptedPCMPaths(
            micPath: resolvedMic,
            systemPath: resolvedSystem,
            tempFiles: tempFiles
        )
    }
}

File: windows/PabloCompanion/ViewModels/TranscriptionViewModel.cs
using AudioCapture.Storage;
using CommunityToolkit.Mvvm.ComponentModel;
using PabloCompanion.Models;
using PabloCompanion.Services;

namespace PabloCompanion.ViewModels;

/// <summary>
/// Orchestrates cloud-based transcription: uploads encrypted audio sidecars
/// to the Pablo backend, which produces the transcript + SOAP note server-
/// side. Mirrors <c>TranscriptionViewModel.swift</c> on macOS.
///
/// Flow:
///   1. <see cref="UploadAudioAsync"/> — decrypts PCM sidecars (if encrypted)
///      to temp files, POSTs them to <c>/api/sessions/{id}/upload-audio</c>.
///   2. On a 2xx the backend owns the audio, so the local recording is deleted
///      (<see cref="RecordingCleaner"/>) — it is PHI, and it is large.
///   3. On failure the session stays in <see cref="PendingTranscriptionStore"/>
///      so it can be retried, and its audio stays on disk.
///   4. <see cref="ResumePendingUploadsAsync"/> runs on app launch with
///      exponential backoff; survives sign-out / sign-in because the pending
///      entries are AES-GCM encrypted with the device key and carry the audio
///      paths inline.
/// </summary>
public partial class TranscriptionViewModel : ObservableObject
{
    private readonly SessionRecordingStore _recordingStore;
    private readonly PendingTranscriptionStore _pendingStore;
    private readonly APIClient _apiClient;
    private readonly CredentialManager _credentials;
    private readonly RecordingCleaner _cleaner;

    // Exponential backoff — matches macOS (PendingTranscriptStore behavior).
    private const int BaseBackoffSeconds = 300;    // 5 minutes
    private const int MaxBackoffSeconds = 14400;   // 4 hours
    private const int MaxAutoRetries = 10;

    [ObservableProperty]
    public partial TranscriptionState State { get; set; } = TranscriptionState.Idle;

    [ObservableProperty]
    public partial double Progress { get; set; }

    [ObservableProperty]
    public partial string? ProgressMessage { get; set; }

    [ObservableProperty]
    public partial string? ErrorMessage { get; set; }

    [ObservableProperty]
    public partial string? ActiveSessionId { get; set; }

    [ObservableProperty]
    public partial int PendingUploadCount { get; set; }

    public TranscriptionViewModel(
        SessionRecordingStore recordingStore,
        PendingTranscriptionStore pendingStore,
        APIClient apiClient,
        CredentialManager credentials,
        RecordingCleaner cleaner)
    {
        _recordingStore = recordingStore;
        _pendingStore = pendingStore;
        _apiClient = apiClient;
        _credentials = credentials;
        _cleaner = cleaner;

        PendingUploadCount = _pendingStore.GetAll().Length;
    }

    /// <summary>
    /// Enqueue a session for cloud transcription and kick off an immediate
    /// upload attempt. Called by <c>SessionViewModel.EndSessionAsync</c>.
    /// </summary>
    public async Task UploadAudioAsync(string sessionId)
    {
        if (State == TranscriptionState.Uploading) return;

        var recording = _recordingStore.Get(sessionId);
        if (recording?.MicPcmFilePath == null)
        {
            ErrorMessage = "No recording found for this session.";
            State = TranscriptionState.Error;
            return;
        }

        // Persist paths so retries survive sign-out (which wipes the recording store).
        _pendingStore.Add(
            sessionId: sessionId,
            micPath: recording.MicPcmFilePath,
            systemPath: recording.SystemPcmFilePath,
            isEncrypted: recording.IsEncrypted);
        PendingUploadCount = _pendingStore.GetAll().Length;

        ActiveSessionId = sessionId;

        // Cheap read-only liveness probe before moving the audio. If the
        // server-side session has already idled out, the upload can only 401 —
        // surface the re-auth flow now (VerifySessionAliveAsync raises it) and
        // leave the entry queued. It drains via the retry loop after sign-in.
        if (!await _apiClient.VerifySessionAliveAsync())
        {
            State = TranscriptionState.PendingUpload;
            ProgressMessage = "Session expired — sign in to resume the upload";
            App.Log($"Skipping audio upload for session={sessionId}: server session is no longer active");
            return;
        }

        State = TranscriptionState.Uploading;
        Progress = 0.1;
        ProgressMessage = "Uploading audio to Pablo...";
        ErrorMessage = null;

        var uploaded = await UploadFromPendingAsync(new PendingTranscription(
            SessionId: sessionId,
            MicPath: recording.MicPcmFilePath,
            SystemPath: recording.SystemPcmFilePath,
            IsEncrypted: recording.IsEncrypted,
            CreatedAt: DateTime.UtcNow,
            RetryCount: 0));

        if (uploaded)
        {
            State = TranscriptionState.Complete;
            Progress = 1.0;
            ProgressMessage = "Upload complete";
        }
        else
        {
            State = TranscriptionState.PendingUpload;
            ProgressMessage = "Upload failed — will retry later";
        }
    }

    /// <summary>
    /// Retry queued uploads with exponential backoff. Invoked on app launch.
    /// </summary>
    public async Task ResumePendingUploadsAsync()
    {
        var pending = _pendingStore.GetAll();
        PendingUploadCount = pending.Length;
        if (pending.Length == 0) return;

        App.Log($"ResumePendingUploads: {pending.Length} item(s) in queue");

        foreach (var item in pending)
        {
            if (item.RetryCount >= MaxAutoRetries)
            {
                App.Log($"  skip session={item.SessionId} retry={item.RetryCount} (max retries exhausted)");
                continue;
            }
            if (item.RetryCount > 0)
            {
                var backoffSeconds = Math.Min(
                    BaseBackoffSeconds * Math.Pow(2, item.RetryCount - 1),
                    MaxBackoffSeconds);
                if ((DateTime.UtcNow - item.CreatedAt).TotalSeconds < backoffSeconds)
                {
                    App.Log($"  skip session={item.SessionId} retry={item.RetryCount} (backoff)");
                    continue;
                }
            }

            await UploadFromPendingAsync(item);
        }
    }

    /// <summary>
    /// Manual "Retry now" — ignores backoff, tries every pending item once.
    /// </summary>
    public async Task ForceRetryPendingUploadsAsync()
    {
        foreach (var item in _pendingStore.GetAll())
        {
            await UploadFromPendingAsync(item);
        }
    }

    /// <summary>
    /// Current transcription state for a specific session, used by UI badges.
    /// </summary>
    public TranscriptionState GetSessionTranscriptionState(string sessionId)
    {
        return _pendingStore.Get(sessionId) != null
            ? TranscriptionState.PendingUpload
            : TranscriptionState.Complete;
    }

    /// <summary>Signed-out cleanup. PendingTranscriptions stay intact so
    /// uploads can resume on sign-in; they are already encrypted at rest.
    /// </summary>
    public void ClearAllData()
    {
        State = TranscriptionState.Idle;
        Progress = 0;
        ProgressMessage = null;
        ErrorMessage = null;
        ActiveSessionId = null;
    }

    // --- internal ---

    private async Task<bool> UploadFromPendingAsync(PendingTranscription item)
    {
        try
        {
            AesGcmEncryptor? encryptor = null;
            if (item.IsEncrypted)
            {
                var keyBytes = _credentials.GetOrCreateUserEncryptionKey();
                if (keyBytes != null)
                    encryptor = new AesGcmEncryptor(keyBytes, "device-key");
            }

            // Decryption runs inside the try on purpose: a missing key throws, and that
            // has to land on the retry path below rather than escape. Neither caller of
            // this method wraps it, so an escaping throw would abort the whole resume
            // pass and strand every remaining item behind this one.
            using var mic = await PcmDecryptor.PrepareForUploadAsync(item.MicPath, encryptor);
            using DecryptedPcm? sys = item.SystemPath != null && File.Exists(item.SystemPath)
                ? await PcmDecryptor.PrepareForUploadAsync(item.SystemPath, encryptor)
                : null;

            // Self-healing upload: a session whose status PATCH never landed is
            // still "recording" server-side and rejects its audio with 400
            // INVALID_STATUS. The core client PATCHes it to recording_complete and
            // retries once; anything else lands here and takes the retry/backoff path.
            await _apiClient.UploadAudioWithSelfHealAsync(item.SessionId, mic.Path, sys?.Path);
            _pendingStore.Remove(item.SessionId);
            PendingUploadCount = _pendingStore.GetAll().Length;
            App.Log($"  upload OK session={item.SessionId} retry={item.RetryCount}");

            // Only now is the audio genuinely safe to destroy: this call returned
            // 2xx, which is the backend confirming it has the recording and has
            // queued it for transcription. Nothing earlier in the flow proves that.
            //
            // Caught here rather than falling through to the retry path below: a
            // delete failure must never turn a confirmed upload back into a retry.
            // The files simply stay, the next launch re-adopts them, and the
            // re-upload takes a bounded INVALID_STATUS rejection. The cleaner
            // already swallows its own errors — this guard is what keeps that from
            // being load-bearing.
            try
            {
                _cleaner.DeleteSession(item.SessionId);
            }
            catch (Exception ex)
            {
                App.LogException("TranscriptionViewModel.DeleteAfterUpload", ex);
            }

            return true;
        }
        catch (Exception ex)
        {
            _pendingStore.IncrementRetry(item.SessionId);
            ErrorMessage = $"Audio upload failed: {ex.Message}";
            App.Log($"  upload FAIL session={item.SessionId} retry={item.RetryCount}->{item.RetryCount + 1} type={ex.GetType().Name}{FormatStatusCode(ex.Message)}");
            return false;
        }
    }

    /// <summary>
    /// Extracts only the HTTP status digits from the upload exception (e.g. "(401)").
    /// Discards the rest of the message — the response body can contain PHI/PII.
    /// </summary>
    private static string FormatStatusCode(string? message)
    {
        if (string.IsNullOrEmpty(message)) return "";
        var match = System.Text.RegularExpressions.Regex.Match(message, @"\((\d{3})\)");
        return match.Success ? $" status={match.Groups[1].Value}" : "";
    }
}
